### DISCUSSION TITLE
API to register side effects in transform

### DISCUSSION THREAD

**Discussion by nishanthkarthik** [11ty/eleventy#4093] [Enhancement Queue]
### Is your feature request related to a problem? Please describe.

How can I use the 11ty API to do this in an `addTransform`?

1. input is a markdown file
2. parse markdown and extract img tags
3. transform markdown to html **and** download images
4. replace the urls in the html with paths to downloaded content

I can do everything with the `addTransform` API except writing to the file system because transforms are supposed to be _pure_ in my understanding. Is there a way to use 11ty API to emit one more output file instead of writing to the file system directly on my own? 

### Describe the solution you'd like

The downside of writing to the file system is that it breaks incremental builds - 11ty wouldn't know when an image changed, so I need to implement a crude version of caching in the transform to ensure I don't download images on each incremental build. If 11ty did support an API like `config.registerResource(url, outputPath)` and allowed customization on cache behavior (file size, last modified header through a HEAD request, always cached), that would be great.

**Comment by zachleat** [comment]
> By default this repo closes enhancement request issues and places them in a queue for folks to upvote. We do this to avoid a large build-up of open issues. We got this idea from the [lodash project](https://twitter.com/samselikoff/status/991395669016436736).

> Don’t forget to upvote the first comment with 👍 to register your vote! You can view [the current enhancement backlog](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+).

**Comment by zachleat** [comment]
I’d shout out the HTML Transform method for the image plugin here: https://www.11ty.dev/docs/plugins/image/#html-transform (but note that this request is more lower level)
